schema/proto.go

```diff
@@ -0,0 +1,386 @@
+func GenerateFieldTag(field reflect.StructField) (tagString string, err error) {
+	tags, err := GetDefinedTypeTag(field.Type)
+	if err != nil {
+		return "", err
+	}
+	tags[ParquetFieldName] = field.Name
+	return getTagStringFromTags(tags), nil
+}
+
+func getTagStringFromTags(tags map[string]string) string {
+	var pairs []string
+	for key, value := range tags {
+		pairs = append(pairs, fmt.Sprintf("%s=%s", key, value))
+	}
+	return strings.Join(pairs, ", ")
+}
+
+func getMapTag(tp reflect.Type) (tags map[string]string, err error) {
+	tags = make(map[string]string)
+	tags[ParquetType] = ParquetTypeMap
+	tags[ParquetConvertedType] = ParquetTypeMap
+	if IsPrimitiveOrPointerGoKind(tp.Key().Kind()) {
+		elementTags, elementErr := GetDefinedTypeTag(tp.Key())
+		if elementErr != nil {
+			return nil, elementErr
+		}
+		if elementType, exists := elementTags[ParquetType]; exists && IsPrimitiveParquetType(elementType) {
+			tags[ParquetKeyType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetKeyConvertedType] = elementConvertedType
+			}
+		}
+	}
+	if IsPrimitiveOrPointerGoKind(tp.Elem().Kind()) {
+		elementTags, elementErr := GetDefinedTypeTag(tp.Elem())
+		if elementErr != nil {
+			return nil, elementErr
+		}
+		if elementType, exists := elementTags[ParquetType]; exists && IsPrimitiveParquetType(elementType) {
+			tags[ParquetValueType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetValueConvertedType] = elementConvertedType
+			}
+		}
+	}
+	return
+}
+
+func getListTag(tp reflect.Type) (tags map[string]string, err error) {
+	tags = make(map[string]string)
+	tags[ParquetType] = ParquetTypeList
+	tags[ParquetConvertedType] = ParquetTypeList
+	if IsPrimitiveOrPointerGoKind(tp.Elem().Kind()) {
+		elementTags, elementErr := GetDefinedTypeTag(tp.Elem())
+		if elementErr != nil {
+			return nil, elementErr
+		}
+		if elementType, exists := elementTags[ParquetType]; exists && IsPrimitiveParquetType(elementType) {
+			tags[ParquetValueType] = elementType
+			if elementConvertedType, exists := elementTags[ParquetConvertedType]; exists {
+				tags[ParquetValueConvertedType] = elementConvertedType
+			}
+		}
+	}
+	return
+}
+
+// Speical handling for the timestamp schema to convert it as millis as one int64 number instead of keeping it as struct.
+func generateSchemaTimestamp(item *Item) (*parquet.SchemaElement, *common.Tag, error) {
+	item.Info.Type = ParquetTypeInt64
+	item.Info.ConvertedType = ConvertedTypeTimestampMills
+	schema, err := common.NewSchemaElementFromTagMap(item.Info)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate schema for timestamp: %v", err)
+	}
+	newInfo := common.NewTag()
+	common.DeepCopy(item.Info, newInfo)
+	return schema, newInfo, nil
+}
+
+func NewSchemaHandlerFromProtoStruct(obj interface{}) (sh *SchemaHandler, err error) {
```

Comment:
Given the conditionality, leveraging stack and other nuances involved in this method, I'd highly recommend a small paragraph description this method's behavior and what's trying to accomplish. Thanks!